File: pyweather/usage_report.py
"""Part 6 — Reading your own usage log.

Run: uv run usage

Reads usage.csv and tells you what you've spent. Makes NO API calls, so it
is free and you can run it as often as you like.

This is the Console dashboard you don't have. It also proves a point worth
internalizing: once you write facts down in a boring format, you can answer
questions nobody planned for. The file is just rows and columns — this
script is one way to read it, and Excel is another.

Against src/usage-report.ts, the parsing itself is the lesson. That file
hand-rolls `splitCsvLine`, a small state machine for quoted fields and ""
escapes, because Node has no CSV parser in its standard library. Python does
— `csv.reader` handles quoting and escaping for you. Two languages, same
file format, and one of them needed forty lines nobody asked to write.

We still look columns up BY NAME rather than by position, exactly like the
TypeScript's `headers.indexOf(name)` — the ledger's column order isn't a
contract either language should hardcode.
"""

import csv
from dataclasses import dataclass
from typing import Callable

from .usage import LEDGER


@dataclass
class Row:
    run_id: str
    script: str
    model: str
    input_tokens: int
    cache_read: int
    cache_write: int
    thinking_tokens: int
    output_tokens: int
    context_tokens: int
    cost_usd: float


def _to_int(value: str) -> int:
    # Same behavior as TypeScript's Number(''): an empty or missing cell
    # parses as 0, not an error.
    return int(value) if value else 0


def _to_float(value: str) -> float:
    return float(value) if value else 0.0


def main() -> None:
    if not LEDGER.exists():
        print(f"No {LEDGER.name} yet. Run any script that calls Claude, then try again.")
        return

    # encoding="utf-8-sig" strips the byte-order mark usage.py writes for
    # Excel's sake, the same job as src/usage-report.ts's
    # `.replace(/^﻿/, '')`. newline="" hands the raw file to csv.reader
    # so it — not Python's text-mode newline translation — decides where
    # rows end.
    with LEDGER.open("r", encoding="utf-8-sig", newline="") as f:
        lines = list(csv.reader(f))

    # A file that exists but is empty is a real state: a run killed between
    # creating usage.csv and writing the header. src/usage-report.ts survives
    # it, so this must too — the two reports have to agree on every input,
    # not just the happy one.
    if not lines:
        print(f"No rows in {LEDGER.name} yet. Run any script that calls Claude, then try again.")
        return

    headers = lines[0]

    def get(cells: list[str], name: str) -> str:
        # Missing column, or a row shorter than the header — the last row of a
        # ledger whose append was interrupted. Both read as empty rather than
        # raising, matching `?? ''` in src/usage-report.ts.
        if name not in headers:
            return ""
        i = headers.index(name)
        return cells[i] if i < len(cells) else ""

    rows = [
        Row(
            run_id=get(cells, "run_id"),
            script=get(cells, "script"),
            model=get(cells, "model"),
            input_tokens=_to_int(get(cells, "input_tokens")),
            cache_read=_to_int(get(cells, "cache_read")),
            cache_write=_to_int(get(cells, "cache_write")),
            thinking_tokens=_to_int(get(cells, "thinking_tokens")),
            output_tokens=_to_int(get(cells, "output_tokens")),
            context_tokens=_to_int(get(cells, "context_tokens")),
            cost_usd=_to_float(get(cells, "cost_usd")),
        )
        for cells in lines[1:]
        if cells  # csv.reader yields [] for a trailing blank line
    ]

    def total(pick: Callable[[Row], float]) -> float:
        return sum(pick(r) for r in rows)

    def money(dollars: float) -> str:
        return f"${dollars:.4f}"

    print(f"\n=== {len(rows)} calls in {LEDGER.name} ===\n")
    print(f"Total spend      {money(total(lambda r: r.cost_usd))}")
    print(f"Input tokens     {total(lambda r: r.input_tokens):,.0f} uncached")
    print(f"Output tokens    {total(lambda r: r.output_tokens):,.0f}")

    thinking = total(lambda r: r.thinking_tokens)
    if thinking > 0:
        output = total(lambda r: r.output_tokens)
        share = f"{(thinking / output * 100):.1f}"
        print(
            f"  of which        {thinking:,.0f} were thinking ({share}%) — "
            "reasoning you paid for and never saw"
        )

    # --- By model -----------------------------------------------------------
    print("\n--- by model ---")
    for model in dict.fromkeys(r.model for r in rows):
        mine = [r for r in rows if r.model == model]
        cost = sum(r.cost_usd for r in mine)
        print(f"{model.ljust(28)} {str(len(mine)).rjust(4)} calls  {money(cost).rjust(10)}")

    # --- By session -----------------------------------------------------------
    # This is the one that answers "what did that conversation cost me?"
    print("\n--- by session (one run_id per program start) ---")
    for run_id in dict.fromkeys(r.run_id for r in rows):
        mine = [r for r in rows if r.run_id == run_id]
        cost = sum(r.cost_usd for r in mine)
        scripts = ", ".join(dict.fromkeys(r.script for r in mine))
        print(f"{run_id}  {str(len(mine)).rjust(4)} calls  {money(cost).rjust(10)}  {scripts}")

    # --- Caching --------------------------------------------------------------
    cache_read = total(lambda r: r.cache_read)
    cache_write = total(lambda r: r.cache_write)

    print("\n--- caching ---")
    if cache_read == 0 and cache_write == 0:
        print("No cache activity yet. Either caching is off, or every prompt")
        print("was under the minimum (1,024 tokens on Sonnet 5). See Part 11.")
    else:
        print(f"Written to cache  {cache_write:,.0f} tokens (billed at 1.25x)")
        print(f"Read from cache   {cache_read:,.0f} tokens (billed at 0.1x)")
        print(
            f"Those reads cost you {(cache_read * 0.1):,.0f} tokens' worth "
            f"instead of {cache_read:,.0f}."
        )

    print(f"\nSame numbers, no code: open {LEDGER.name} in Excel and sum the cost_usd column.\n")

File: src/usage-report.ts
// Part 6 — Reading your own usage log.
//
// Run: npm run usage
//
// Reads usage.csv and tells you what you've spent. Makes NO API calls, so it
// is free and you can run it as often as you like.
//
// This is the Console dashboard you don't have. It also proves a point worth
// internalizing: once you write facts down in a boring format, you can answer
// questions nobody planned for. The file is just rows and columns — this
// script is one way to read it, and Excel is another.

import { existsSync, readFileSync } from 'node:fs';

const FILE = 'usage.csv';

if (!existsSync(FILE)) {
  console.log(`No ${FILE} yet. Run any script that calls Claude, then try again.`);
  process.exit(0);
}

interface Row {
  run_id: string;
  script: string;
  model: string;
  input_tokens: number;
  cache_read: number;
  cache_write: number;
  thinking_tokens: number;
  output_tokens: number;
  context_tokens: number;
  cost_usd: number;
}

/** Splits one CSV line, respecting "quoted, fields" and "" escapes. */
function splitCsvLine(line: string): string[] {
  const cells: string[] = [];
  let cell = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];

    if (inQuotes) {
      if (char === '"' && line[i + 1] === '"') {
        cell += '"';
        i++; // skip the second quote of the pair
      } else if (char === '"') {
        inQuotes = false;
      } else {
        cell += char;
      }
    } else if (char === '"') {
      inQuotes = true;
    } else if (char === ',') {
      cells.push(cell);
      cell = '';
    } else {
      cell += char;
    }
  }

  cells.push(cell);
  return cells;
}

// .replace(/^﻿/, '') strips the byte-order mark we wrote for Excel's sake.
const lines = readFileSync(FILE, 'utf8').replace(/^﻿/, '').trim().split('\n');
const headers = splitCsvLine(lines[0]!);

const rows: Row[] = lines.slice(1).map((line) => {
  const cells = splitCsvLine(line);
  const get = (name: string) => cells[headers.indexOf(name)] ?? '';
  return {
    run_id: get('run_id'),
    script: get('script'),
    model: get('model'),
    input_tokens: Number(get('input_tokens')),
    cache_read: Number(get('cache_read')),
    cache_write: Number(get('cache_write')),
    thinking_tokens: Number(get('thinking_tokens')),
    output_tokens: Number(get('output_tokens')),
    context_tokens: Number(get('context_tokens')),
    cost_usd: Number(get('cost_usd')),
  };
});

const sum = (pick: (r: Row) => number) => rows.reduce((total, r) => total + pick(r), 0);
const money = (dollars: number) => `$${dollars.toFixed(4)}`;

console.log(`\n=== ${rows.length} calls in ${FILE} ===\n`);
console.log(`Total spend      ${money(sum((r) => r.cost_usd))}`);
console.log(`Input tokens     ${sum((r) => r.input_tokens).toLocaleString()} uncached`);
console.log(`Output tokens    ${sum((r) => r.output_tokens).toLocaleString()}`);

const thinking = sum((r) => r.thinking_tokens);
if (thinking > 0) {
  const output = sum((r) => r.output_tokens);
  const share = ((thinking / output) * 100).toFixed(1);
  console.log(
    `  of which        ${thinking.toLocaleString()} were thinking (${share}%) — ` +
      `reasoning you paid for and never saw`,
  );
}

// --- By model ---------------------------------------------------------------
console.log('\n--- by model ---');
for (const model of [...new Set(rows.map((r) => r.model))]) {
  const mine = rows.filter((r) => r.model === model);
  const cost = mine.reduce((total, r) => total + r.cost_usd, 0);
  console.log(
    `${model.padEnd(28)} ${String(mine.length).padStart(4)} calls  ${money(cost).padStart(10)}`,
  );
}

// --- By session -------------------------------------------------------------
// This is the one that answers "what did that conversation cost me?"
console.log('\n--- by session (one run_id per program start) ---');
for (const runId of [...new Set(rows.map((r) => r.run_id))]) {
  const mine = rows.filter((r) => r.run_id === runId);
  const cost = mine.reduce((total, r) => total + r.cost_usd, 0);
  const scripts = [...new Set(mine.map((r) => r.script))].join(', ');
  console.log(
    `${runId}  ${String(mine.length).padStart(4)} calls  ${money(cost).padStart(10)}  ${scripts}`,
  );
}

// --- Caching ----------------------------------------------------------------
const cacheRead = sum((r) => r.cache_read);
const cacheWrite = sum((r) => r.cache_write);
const anyGrok = rows.some((r) => r.model.startsWith('grok'));

console.log('\n--- caching ---');
if (anyGrok) {
  // Grok (or mixed) rows: Claude's 0.1× / 1,024 / 1.25× copy is the wrong
  // story. Grok caches automatically, writes cost nothing extra, and the
  // dollar savings are already in cost_usd.
  console.log(`Written to cache  ${cacheWrite.toLocaleString()} tokens`);
  console.log(`Read from cache   ${cacheRead.toLocaleString()} tokens`);
  console.log('Grok cache_write is always 0 — there is no separate write premium.');
  console.log('Savings are already in cost_usd; this report does not re-derive them.');
} else if (cacheRead === 0 && cacheWrite === 0) {
  console.log('No cache activity yet. Either caching is off, or every prompt');
  console.log('was under the minimum (1,024 tokens on Sonnet 5). See Part 11.');
} else {
  console.log(`Written to cache  ${cacheWrite.toLocaleString()} tokens (billed at 1.25x)`);
  console.log(`Read from cache   ${cacheRead.toLocaleString()} tokens (billed at 0.1x)`);
  console.log(
    `Those reads cost you ${(cacheRead * 0.1).toLocaleString()} tokens' worth ` +
      `instead of ${cacheRead.toLocaleString()}.`,
  );
}

console.log(
  `\nSame numbers, no code: open ${FILE} in Excel and sum the cost_usd column.\n`,
);
